snuba/query/matcher.py

```diff
@@ -0,0 +1,218 @@
+
+from snuba.query.expressions import Column as ColumnExpr
+from snuba.query.expressions import Expression, FunctionCall
+
+TNode = TypeVar("TNode")
+
+
+@dataclass(frozen=True)
+class MatchResult:
+    """
+    Contains the parameters found in an expression by a Pattern.
+    Parameters are the equivalent to groups in regular expressions, they
+    are named parts of the expression we are matching that are identified
+    when a valid match is found.
+    """
+
+    expressions: Mapping[str, Expression] = field(default_factory=dict)
+    strings: Mapping[str, Optional[str]] = field(default_factory=dict)
+    ints: Mapping[str, Optional[int]] = field(default_factory=dict)
+    floats: Mapping[str, Optional[float]] = field(default_factory=dict)
+    dates: Mapping[str, Optional[date]] = field(default_factory=dict)
+    datetimes: Mapping[str, Optional[datetime]] = field(default_factory=dict)
+
+    def merge(self, values: MatchResult) -> MatchResult:
+        return MatchResult(
+            expressions={**self.expressions, **values.expressions},
+            strings={**self.strings, **values.strings},
+            ints={**self.ints, **values.ints},
+            floats={**self.floats, **values.floats},
+            dates={**self.dates, **values.dates},
+            datetimes={**self.datetimes, **values.datetimes},
+        )
+
+
+class Pattern(ABC, Generic[TNode]):
+    """
+    Tries to match a given node (like an AST Expression) with the rules provided
+    when instantiating this class. This is meant to work like a regular expression
+    but for AST expressions.
+
+    When a node is provided to this method, if this method return an instance of
+    MatchResult it means the node fully matches the pattern.
+    If the pattern defines named parameters, their value is returned as part of
+    the MatchResult object returned.
+    More on how to define parameters in the Param class.
+
+    This is not supposed to ever throw as long as the node is properly is a valid
+    data structure, even if it does not match the pattern.
+    """
+
+    @abstractmethod
+    def match(self, node: TNode) -> Optional[MatchResult]:
+        """
+        Returns a MatchResult if the node provided matches this pattern
+        otherwise it returns None.
+        """
+        raise NotImplementedError
+
+
+@dataclass(frozen=True)
+class Param(Pattern[TNode], Generic[TNode]):
+    """
+    Defines a named parameter in a Pattern. When matching the overall Pattern,
+    if the Pattern nested in this object matches, the corresponding input node
+    will be given a name and returned as part of the MatchResult return value.
+    This is meant to represent what a group is in a regex.
+
+    Params can be nested, and they can wrap any valid Pattern.
+    """
+
+    name: str
+    pattern: Pattern[TNode]
+
+    def match(self, node: TNode) -> Optional[MatchResult]:
+        result = self.pattern.match(node)
+        if not result:
```

Comment:
```suggestion
        if result is None:
```

[PEP 8 recommends this to avoid bugs](https://www.python.org/dev/peps/pep-0008/#id51) and I think this would be more explicit as well, especially if you consider a `MatchResult` with no named parameter matches as an empty collection. (I know it's not _literally_ that, but it's an easy misinterpretation to make when skimming this.)

(This applies multiple times in this file below, I'm not going to note each of them.)

Reply:
I did not use is None because MatchResult does not match an empty collection, but make sense. Will change.